File: C/2/9.c
#include <stdio.h>

/**
 * A function that counts the number of 1-bits in its integer argument.
 * 
 * @param unsigned x Number to have its 1-bits counted
 * 
 * @return int       Number of 1-bits
 */
int bitcount(unsigned x)
{
    int b = 0;

    /**
     * Q: Why operation (x &= (x-1)) deletes the rightmost 1-bit in x?
     * 
     * A: Because subtracting 1 turns off the rightmost bit in x, and applying
     * bitwise and guarantees that we are not adding new bits to x (after all,
     * we want to count how many there are).
     * 
     * We keep on counting while x is greater than zero and the assignment
     * operation returns greater than or equals to zero because the last
     * operation (the one that turns x into 0) is also a deletion of the
     * rightmost bit and needs to be counted (think of x = 8).
     */
    while (x > 0 && (x &= (x-1)) >= 0) {
        b++;
    }
    
    return b;
}

/**
 * A program to test bitcount.
 */
main()
{
    printf("bitcount(255)\t=>\t%u\n", bitcount(255));
    printf("bitcount(56452)\t=>\t%u\n", bitcount(56452));
    printf("bitcount(0)\t=>\t%u\n", bitcount(0));
    printf("bitcount(~0)\t=>\t%u\n", bitcount(~0));
    printf("bitcount(10)\t=>\t%u\n", bitcount(10));
    printf("bitcount(32)\t=>\t%u\n", bitcount(32));
    printf("bitcount(16)\t=>\t%u\n", bitcount(16));
    printf("bitcount(8)\t=>\t%u\n", bitcount(8));
    printf("bitcount(4)\t=>\t%u\n", bitcount(4));
    printf("bitcount(2)\t=>\t%u\n", bitcount(2));
    printf("bitcount(1)\t=>\t%u\n", bitcount(1));
}

File: rustlang/2/9.rs
/// A function that counts the number of 1-bits in its integer argument.
/// 
/// @param  u32 x Number to have its 1-bits counted
/// 
/// @return i32   Number of 1-bits
fn bitcount(x : u32) -> i32
{
    let mut b = 0;

    // We cast x to a signed integer so we can check if
    // the operation was succesful
    let mut n : i32 = x as i32;

    while n > 0 {
        // We make the operation here because in Rust, different from C,
        // the result of the assignment is not returned, so we were not able to
        // do if (n &= n-1) >= 0
        n &= n-1;

        // If n is not declared as i32, the compiler will throw an warning
        // for this comparison (because u32 is never less than 0).
        if n >= 0 {
            b += 1;
        }
    }

    b
}

/// A program to test bitcount.
fn main()
{
    print!("bitcount(255)\t=>\t{}\n", bitcount(255));
    print!("bitcount(56452)\t=>\t{}\n", bitcount(56452));
    print!("bitcount(0)\t=>\t{}\n", bitcount(0));
    print!("bitcount(!0)\t=>\t{}\n", bitcount(!0));
    print!("bitcount(10)\t=>\t{}\n", bitcount(10));
    print!("bitcount(32)\t=>\t{}\n", bitcount(32));
    print!("bitcount(16)\t=>\t{}\n", bitcount(16));
    print!("bitcount(8)\t=>\t{}\n", bitcount(8));
    print!("bitcount(4)\t=>\t{}\n", bitcount(4));
    print!("bitcount(2)\t=>\t{}\n", bitcount(2));
    print!("bitcount(1)\t=>\t{}\n", bitcount(1));
}